File: js/src/unit_widgets.ts
import { ComboboxModel, ComboboxView,
       DropdownModel, DropdownView,
       TextModel, TextView,
       TextareaModel, TextareaView,
       ToggleButtonModel, ToggleButtonView } from '@jupyter-widgets/controls';
import { MODULE_NAME, MODULE_VERSION } from './version';

export
class ComboboxUnitModel extends ComboboxModel {
    defaults() {
        return {...super.defaults(),
	_model_name: 'ComboboxUnitModel',
	_model_module: MODULE_NAME,
	_model_module_version: MODULE_VERSION,
	_view_name: 'ComboboxUnitView',
	_view_module: MODULE_NAME,
	_view_module_version: MODULE_VERSION,
        _focus: null,
	_tooltip: null,
	tabindex: null,
        };
    }
}

export
class DropdownUnitModel extends DropdownModel {
    defaults() {
        return {...super.defaults(),
	_model_name: 'DropdownUnitModel',
	_model_module: MODULE_NAME,
	_model_module_version: MODULE_VERSION,
	_view_name: 'DropdownUnitView',
	_view_module: MODULE_NAME,
	_view_module_version: MODULE_VERSION,
        _focus: null,
	_tooltip: null,
	tabindex: null,
        };
    }
}

export
class TextUnitModel extends TextModel {
    defaults() {
        return {...super.defaults(),
	_model_name: 'TextUnitModel',
	_model_module: MODULE_NAME,
	_model_module_version: MODULE_VERSION,
	_view_name: 'TextUnitView',
	_view_module: MODULE_NAME,
	_view_module_version: MODULE_VERSION,
        _focus: null,
	_tooltip: null,
	tabindex: null,
        };
    }
}

export
class TextareaUnitModel extends TextareaModel {
    defaults() {
        return {...super.defaults(),
	_model_name: 'TextareaUnitModel',
	_model_module: MODULE_NAME,
	_model_module_version: MODULE_VERSION,
	_view_name: 'TextareaUnitView',
	_view_module: MODULE_NAME,
	_view_module_version: MODULE_VERSION,
        _focus: null,
	_tooltip: null,
	tabindex: null,
        };
    }
}

export
class ToggleButtonUnitModel extends ToggleButtonModel {
    defaults() {
        return {...super.defaults(),
	_model_name: 'ToggleButtonUnitModel',
	_model_module: MODULE_NAME,
	_model_module_version: MODULE_VERSION,
	_view_name: 'ToggleButtonUnitView',
	_view_module: MODULE_NAME,
	_view_module_version: MODULE_VERSION,
        _focus: null,
	_tooltip: null,
	tabindex: null,
        };
    }
}

export
class ComboboxUnitView extends ComboboxView {
    render() {
        super.render();
        this.update_tabindex();
        this.update_tooltip();
        this.update_title();
        this.update_focus();
        this.model.on('change:tabindex', this.update_tabindex, this);
        this.model.on('change:_tooltip', this.update_title, this);
        this.model.on('change:description_tooltip', this.update_tooltip, this);
        this.model.on('change:_focus', this.update_focus, this);
    }

    update_tabindex() {
        let tabindex = this.model.get('tabindex')
        if (tabindex) this.textbox.setAttribute('tabindex', tabindex);
	else this.textbox.removeAttribute('tabindex');
    }

    update_tooltip() {
	this.model.set('_tooltip', this.model.get('description_tooltip'));
    }

    update_title() {
        let title = this.model.get('_tooltip');
        if (title) this.textbox.setAttribute('title', title);
	else this.textbox.removeAttribute('title');
    }

    update_focus() {
        let focus = this.model.get('_focus');
	if (!focus) return;
	if (focus == 'on') { this.textbox.focus(); }
	else if (focus == 'off') { this.textbox.blur(); }
    }
};

export
class DropdownUnitView extends DropdownView {
    render() {
        super.render();
        this.update_tabindex();
        this.update_tooltip();
        this.update_title();
        this.update_focus();
        this.model.on('change:tabindex', this.update_tabindex, this);
        this.model.on('change:_tooltip', this.update_title, this);
        this.model.on('change:description_tooltip', this.update_tooltip, this);
        this.model.on('change:_focus', this.update_focus, this);
    }

    update_tabindex() {
        let tabindex = this.model.get('tabindex')
        if (tabindex) this.listbox.setAttribute('tabindex', tabindex);
	else this.listbox.removeAttribute('tabindex');
    }

    update_tooltip() {
	this.model.set('_tooltip', this.model.get('description_tooltip'));
    }

    update_title() {
        var title = this.model.get('description_tooltip');
        if (!title) this.listbox.removeAttribute('title');
	else this.listbox.setAttribute('title', title);
    }

    update_focus() {
        let focus = this.model.get('_focus');
	if (!focus) return;
	if (focus == 'on') { this.listbox.focus(); }
	else if (focus == 'off') { this.listbox.blur(); }
    }
};

export
class TextUnitView extends TextView {
    render() {
        super.render();
        this.update_tabindex();
        this.update_tooltip();
        this.update_title();
        this.update_focus();
        this.model.on('change:tabindex', this.update_tabindex, this);
        this.model.on('change:_tooltip', this.update_title, this);
        this.model.on('change:description_tooltip', this.update_tooltip, this);
        this.model.on('change:_focus', this.update_focus, this);
    }

    update_tabindex() {
        let tabindex = this.model.get('tabindex')
        if (tabindex) this.textbox.setAttribute('tabindex', tabindex);
	else this.textbox.removeAttribute('tabindex');
    }

    update_tooltip() {
	this.model.set('_tooltip', this.model.get('description_tooltip'));
    }

    update_title() {
        let title = this.model.get('_tooltip');
        if (title) this.textbox.setAttribute('title', title);
	else this.textbox.removeAttribute('title');
    }

    update_focus() {
        let focus = this.model.get('_focus');
	if (!focus) return;
	if (focus == 'on') { this.textbox.focus(); }
	else if (focus == 'off') { this.textbox.blur(); }
    }
};

export
class TextareaUnitView extends TextareaView {
    render() {
        super.render();
        this.update_tabindex();
        this.update_tooltip();
        this.update_title();
        this.update_focus();
        this.model.on('change:tabindex', this.update_tabindex, this);
        this.model.on('change:_tooltip', this.update_title, this);
        this.model.on('change:description_tooltip', this.update_tooltip, this);
        this.model.on('change:_focus', this.update_focus, this);
    }

    update_tabindex() {
        let tabindex = this.model.get('tabindex')
        if (tabindex) this.textbox.setAttribute('tabindex', tabindex);
	else this.textbox.removeAttribute('tabindex');
    }

    update_tooltip() {
	this.model.set('_tooltip', this.model.get('description_tooltip'));
    }

    update_title() {
        let title = this.model.get('_tooltip');
        if (title) this.textbox.setAttribute('title', title);
	else this.textbox.removeAttribute('title');
    }

    update_focus() {
        let focus = this.model.get('_focus');
	if (!focus) return;
	if (focus == 'on') { this.textbox.focus(); }
	else if (focus == 'off') { this.textbox.blur(); }
    }
};

export
class ToggleButtonUnitView extends ToggleButtonView {
    render() {
        super.render();
        this.update_tabindex();
        this.update_tooltip();
        this.update_title();
        this.update_focus();
        this.model.on('change:tabindex', this.update_tabindex, this);
        this.model.on('change:_tooltip', this.update_title, this);
        this.model.on('change:tooltip', this.update_tooltip, this);
        this.model.on('change:description_tooltip', this.update_tooltip, this);
        this.model.on('change:_focus', this.update_focus, this);
    }

    update_tabindex() {
        let tabindex = this.model.get('tabindex')
        if (tabindex) this.el.setAttribute('tabindex', tabindex);
	else this.el.removeAttribute('tabindex');
    }

    update_tooltip() {
        if (this.model.get('tooltip')) this.model.set('_tooltip', this.model.get('description_tooltip'));
	else this.model.set('_tooltip', this.model.get('description_tooltip'));
    }

    update_title() {
        let title = this.model.get('_tooltip');
        if (title) this.el.setAttribute('title', title);
	else this.el.removeAttribute('title');
    }

    update_focus() {
        let focus = this.model.get('_focus');
	if (!focus) return;
	if (focus == 'on') { this.el.focus(); }
	else if (focus == 'off') { this.el.blur(); }
    }
};

File: unit_widgets/unit_widgets.py
# -*- coding: utf-8 -*-
r"""
IPywidgets with simple additional features to serve as units to larger widgets.

AUTHORS ::

    Odile Bénassy, Nicolas Thiéry

"""
from traitlets import HasTraits, Int, Unicode
from ipywidgets import Button, Combobox, Dropdown, HTMLMath, Text, Textarea, ToggleButton, register


class Unit(HasTraits):
    """Additional features to an ipywidgets widget."""
    _focus = Unicode().tag(sync=True)
    _tooltip = Unicode('').tag(sync=True) # set '' as default value
    tabindex = Int().tag(sync=True)

    def set_tooltip(self, s=''):
        self._tooltip = s

    def focus(self):
        self._focus = ''
        self._focus = 'on'

    def blur(self):
        self._focus = ''
        self._focus = 'off'

    def set_tabindex(self, i=0):
        self.tabindex = i

    def allow_focus(self):
        self.set_tabindex(0)

    def disallow_focus(self):
        self.set_tabindex(-1)


@register
class ButtonUnit(Button, Unit):
    """Button with tooltip and focus."""
    _model_name = Unicode('ButtonUnitModel').tag(sync=True)
    _model_module = Unicode('unit-widgets').tag(sync=True)
    _model_module_version = Unicode('^0.7.6').tag(sync=True)
    _view_name = Unicode('ButtonUnitView').tag(sync=True)
    _view_module = Unicode('unit-widgets').tag(sync=True)
    _view_module_version = Unicode('^0.7.6').tag(sync=True)


@register
class ComboboxUnit(Combobox, Unit):
    """Combobox with tooltip and focus."""
    _model_name = Unicode('ComboboxUnitModel').tag(sync=True)
    _model_module = Unicode('unit-widgets').tag(sync=True)
    _model_module_version = Unicode('^0.7.6').tag(sync=True)
    _view_name = Unicode('ComboboxUnitView').tag(sync=True)
    _view_module = Unicode('unit-widgets').tag(sync=True)
    _view_module_version = Unicode('^0.7.6').tag(sync=True)


@register
class DropdownUnit(Dropdown, Unit):
    """Dropdown with tooltip and focus."""
    _model_name = Unicode('DropdownUnitModel').tag(sync=True)
    _model_module = Unicode('unit-widgets').tag(sync=True)
    _model_module_version = Unicode('^0.7.6').tag(sync=True)
    _view_name = Unicode('DropdownUnitView').tag(sync=True)
    _view_module = Unicode('unit-widgets').tag(sync=True)
    _view_module_version = Unicode('^0.7.6').tag(sync=True)


@register
class HTMLMathUnit(HTMLMath, Unit):
    """HTML Math widget with tooltip and focus."""
    _model_name = Unicode('HTMLMathUnitModel').tag(sync=True)
    _model_module = Unicode('unit-widgets').tag(sync=True)
    _model_module_version = Unicode('^0.7.6').tag(sync=True)
    _view_name = Unicode('HTMLMathUnitView').tag(sync=True)
    _view_module = Unicode('unit-widgets').tag(sync=True)
    _view_module_version = Unicode('^0.7.6').tag(sync=True)


@register
class TextUnit(Text, Unit):
    """Input text with tooltip and focus."""
    _model_name = Unicode('TextUnitModel').tag(sync=True)
    _model_module = Unicode('unit-widgets').tag(sync=True)
    _model_module_version = Unicode('^0.7.6').tag(sync=True)
    _view_name = Unicode('TextUnitView').tag(sync=True)
    _view_module = Unicode('unit-widgets').tag(sync=True)
    _view_module_version = Unicode('^0.7.6').tag(sync=True)


@register
class TextareaUnit(Textarea, Unit):
    """Text area with tooltip and focus."""
    _model_name = Unicode('TextareaUnitModel').tag(sync=True)
    _model_module = Unicode('unit-widgets').tag(sync=True)
    _model_module_version = Unicode('^0.7.6').tag(sync=True)
    _view_name = Unicode('TextareaUnitView').tag(sync=True)
    _view_module = Unicode('unit-widgets').tag(sync=True)
    _view_module_version = Unicode('^0.7.6').tag(sync=True)


@register
class ToggleButtonUnit(ToggleButton, Unit):
    """Toggle button with tooltip and focus."""
    _model_name = Unicode('ToggleButtonUnitModel').tag(sync=True)
    _model_module = Unicode('unit-widgets').tag(sync=True)
    _model_module_version = Unicode('^0.7.6').tag(sync=True)
    _view_name = Unicode('ToggleButtonUnitView').tag(sync=True)
    _view_module = Unicode('unit-widgets').tag(sync=True)
    _view_module_version = Unicode('^0.7.6').tag(sync=True)
